src/dtclient/image.go

```diff
@@ -0,0 +1,174 @@
+package dtclient
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+
+	"github.com/pkg/errors"
+)
+
+type LatestImageInfo struct {
+	Source string `json:"source"`
+	Tag    string `json:"tag"`
+}
+
+func (dtc *dynatraceClient) GetLatestOneAgentImage() (*LatestImageInfo, error) {
+	request, err := dtc.createLatestAgentImageRequest()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	response, err := dtc.httpClient.Do(request)
+	if err != nil {
+		log.Info("failed to retrieve latest image")
+		return nil, err
+	}
+
+	latestImageInfo, err := dtc.handleLatestImageResponse(response)
+	if err != nil {
+		log.Info("failed to handle latest image response")
+		return nil, err
+	}
+
+	return latestImageInfo, nil
+}
+
+func (dtc *dynatraceClient) GetLatestCodeModulesImage() (*LatestImageInfo, error) {
+	request, err := dtc.createLatestCodeModulesImageRequest()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	response, err := dtc.httpClient.Do(request)
+	if err != nil {
+		log.Info("failed to retrieve latest image")
+		return nil, err
+	}
+
+	latestImageInfo, err := dtc.handleLatestImageResponse(response)
+	if err != nil {
+		log.Info("failed to handle latest image response")
+		return nil, err
+	}
+
+	return latestImageInfo, nil
+}
+
+func (dtc *dynatraceClient) GetLatestActiveGateImage() (*LatestImageInfo, error) {
+	request, err := dtc.createLatestActiveGateImageRequest()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	response, err := dtc.httpClient.Do(request)
+	if err != nil {
+		log.Info("failed to retrieve latest image")
+		return nil, err
+	}
+
+	latestImageInfo, err := dtc.handleLatestImageResponse(response)
+	if err != nil {
+		log.Info("failed to handle latest image response")
+		return nil, err
+	}
+
+	return latestImageInfo, nil
+}
+
+func (dtc *dynatraceClient) handleLatestImageResponse(response *http.Response) (*LatestImageInfo, error) {
+	defer func() {
+		err := response.Body.Close()
+		if err != nil {
+			log.Error(err, err.Error())
+		}
+	}()
+
+	data, err := dtc.getServerResponseData(response)
+	if err != nil {
+		return nil, dtc.handleErrorResponseFromAPI(data, response.StatusCode)
+	}
+
+	latestImageInfo, err := dtc.readResponseForLatestImage(data)
+	if err != nil {
+		return nil, err
+	}
+	return latestImageInfo, err
+}
+
+func (dtc *dynatraceClient) createLatestAgentImageRequest() (*http.Request, error) {
```

Comment:
you could generalise these `createLatest...ImageRequest`, because the only thing that changes is the URL, so you could pass that in.
```
func (dtc *dynatraceClient) createLatestImageRequest(url string) (*http.Request, error) {
```